### PR TITLE
For CLI testing, generate Python Dependencies Cache

### PR DESCRIPTION
**Pull request by edespino**
This change reduces the numerous and redundant CLI testing pip install
calls to a single invocation. The downloaded python dependencies will
be copied to an intermediate tarball and used in subsequent CLI test
tasks.

**PIPELINE** - applied to pipeline template (gpdb-tpl.yml)

- Introduce the generation of a Python Dependencies cache tarball
  intermediate artifact into the CLI start gate. The cache is platform
  independent and will be used for CentOS and Ubuntu test runs.

- To capture the generation of the python dependencies cache task
  introduced into the CLI start gate, it has been renamed:

  from: 'gate_cli_start'
  to:   'gate_cli_start_gen_python_dependencies_cache'.

- Using Concourse task 'attempts' feature, five attempts to build the
  python dependencies cache will be performed before erroring at the
  CLI gate job.

- Move all CLI Behave jobs (now including the Ubuntu 18 CLI jobs)
  behind the CLI start gate.

- Introduce python dependencies cache input to the following tasks:

    publish_ccp_coverage
    publish_demo_coverage
    combine_coverage

**TASK YMLs**

- Introduce python dependencies cache input to the following task yaml
  files:

    behave_gpdb.yml
    run_behave_on_ccp_cluster.yml
    gpMgmt_check_gpdb.yml

- For the CCP task yaml, copy the python dependencies cache to the
  'mdw' host.

**SCRIPTS**

- common.bash - Introduce a function (extract_python_dependency_cache)
  to be called when installing python requirements for single (DEMO
  cluster) and muti-host (CCP cluster) deployments.

**PIPELINE GENERATION**

- gen_pipeline.py - Update JOBS_THAT_ARE_GATES with renamed CLI start
  gate.

- Regenerate production pipeline - gpdb_master-generated.yml.

**BUGFIX**

- The bin_gpdb_clients_windows_rc resource only applies to the "prod"
  pipeline configuration.

## Dev Pipeline execution examples

**NOTE:** _Unfortunately, the following information is not widely available to the community for review. If you would like additional information, please add a comment and I will do my best to shed light on pipeline execution details._

Below you will find dev pipeline references used to validate the implementation. In addition to dev pipeline links, I have listed the gen_pipeline.py commands used to create the respective dev pipelines.

In the `gen_python_dependencies_cache` task, you will see the `Downloading` message as part of the `pip install` test python dependencies installation (populate $HOME/.cache/pip). All subsequent `pip install` test python dependency invocations will use the python dependencies cache with the `Downloading` references replaced with `Using cached` messages.

- Full dev pipeline - [espino-master-full](https://dev.ci.gpdb.pivotal.io/teams/main/pipelines/espino-master-full) (note Windows environment is not available in dev CI instance and corresponding windows jobs fail)
```
Generate dev pipeline: ./gen_pipeline.py -c full -o espino-master-full.yml
```
- CLI test section - [espino-master-CLI-centos7](https://dev.ci.gpdb.pivotal.io/teams/main/pipelines/espino-master-CLI-centos7)
```
Generate dev pipeline: ./gen_pipeline.py -a CLI -o espino-master-CLI-centos7.yml -O centos7
```
- ICW test section - [espino-master-ICW-centos7](https://dev.ci.gpdb.pivotal.io/teams/main/pipelines/espino-master-ICW-centos7)
```
Generate dev pipeline: ./gen_pipeline.py -a ICW -o espino-master-ICW-centos7.yml -O centos7
```
-  Interconnect test section - [espino-master-Interconnect-centos7](https://dev.ci.gpdb.pivotal.io/teams/main/pipelines/espino-master-Interconnect-centos7)
```
Generate dev pipeline: ./gen_pipeline.py -a Interconnect -o espino-master-Interconnect-centos7.yml -O centos7
```

### CODE REVIEW

File: concourse/pipelines/templates/gpdb-tpl.yml [hunk 1]
```diff
@@ -1136,29 +1148,62 @@ jobs:
...
+      outputs:
+          - name: python_dependencies_cache
+      run:
+        path: bash
+        args:
+          - -exc
+          - |
+            set -eox pipefail
+            source "gpdb_src/concourse/scripts/common.bash"
+            time install_gpdb
```
Comment (tylarb) [inline concourse/pipelines/templates/gpdb-tpl.yml:1187]: installing gpdb should not be necessary for installing the dependencies correct?

File: concourse/scripts/common.bash [hunk 2]
```diff
@@ -69,8 +79,15 @@ function install_python_requirements_on_multi_host() {
...
     local requirements_txt="$1"
 
+    extract_python_dependency_cache
+
     pip install --user -r ${requirements_txt}
     while read -r host; do
+       # if expanded Python dependencies cache exists, copy to other hosts
+       if [ -d $HOME .cache/pip ]; then
+           ssh "$host" mkdir -p .cache
+           scp -r .cache/pip "$host":/tmp/requirements.txt
```
Comment (tylarb) [inline concourse/scripts/common.bash:89]: This is putting the cache in a file called "/tmp/requirements.txt" - don't we want it in `~/.cache/pip` ?

File: concourse/scripts/common.bash [hunk 3]
```diff
@@ -69,8 +79,15 @@ function install_python_requirements_on_multi_host() {
     # the user flag is required for centos 7
     local requirements_txt="$1"
 
+    extract_python_dependency_cache
+
     pip install --user -r ${requirements_txt}
     while read -r host; do
+       # if expanded Python dependencies cache exists, copy to other hosts
+       if [ -d $HOME .cache/pip ]; then
```
Comment (tylarb) [inline concourse/scripts/common.bash:87]: should this be `$HOME/.cache/pip? ` (no space)

**Comment by d** [comment]
TL;DR: -1. Good intention, bad code.

The approach taken here is treating the pip wheel directory more like an _artifact_ than a _cache_. I think the right direction is to leverage Concourse [cache support](https://concourse-ci.org/tasks.html#schema.task.caches)

File: concourse/scripts/common.bash (hunk 2 above)
Comment (edespino) [reply concourse/scripts/common.bash]: @tylarb Thank you.

File: concourse/scripts/common.bash (hunk 3 above)
Comment (edespino) [reply concourse/scripts/common.bash]: @tylarb Thank you.

**Comment by kalensk** [comment]
> I think the right direction is to leverage Concourse [cache support](https://concourse-ci.org/tasks.html#schema.task.caches)

If Im understanding correctly Concourse caches `directories shared between task runs.` Do you think this is this _enough_ to help us solve the overall issue?

**Comment by edespino** [comment]
@d Point taken. I'll review the Concourse cache route. Thank you!.

File: concourse/pipelines/templates/gpdb-tpl.yml (hunk 1 above)
Comment (edespino) [reply concourse/pipelines/templates/gpdb-tpl.yml:1187]: @tylarb I believe the installation of gpdb is required. I'll confirm. Thank you.

Comment (edespino) [reply concourse/pipelines/templates/gpdb-tpl.yml]: I've removed the gpdb install. Thanks.

**Comment by edespino** [comment]
For the time being, I am going to close this PR and migrate to utilizing the Concourse [cache support ](https://concourse-ci.org/tasks.html#schema.task.caches.)
